### PR TITLE
HDPI-5371 text fields accept cross scripting

### PR DESCRIPTION
**Pull request by sdmiddletonSolirius**
### Jira link

See [HDPI-5371](https://tools.hmcts.net/jira/browse/HDPI-5371)

### Change description

- Added the `xss` library https://jsxss.com/en/index.html to sanitise `text`, `textarea`, and `character-count` fields in `validateForm` via `stripHtmlTags` (FilterXSS, empty whitelist, tags stripped). This replaces the earlier denylist that blocked input with validation errors.
- HTML is removed without showing an error so attackers are not told their payload was detected. 
- We only run xss when input looks like HTML (`looksLikeHtml`), so legitimate text such as `1 bedroom is < 2 bedrooms` is not truncated or stored as entities.
- Moved `stripHtmlTags` call to before the `isMissing` check in `validateForm` so that input consisting solely of stripped tags (e.g. '<script>...</script>') correctly fails required field validation rather than being silently accepted.

**BEFORE**
Screenshot 2026-06-01 at 10 31 45

**AFTER**
Screenshot 2026-06-01 at 10 31 54

### Security Vulnerability Assessment ###

**CVE Suppression:** Are there any CVEs present in the codebase (either newly introduced or pre-existing) that are being intentionally suppressed or ignored by this commit?
  * [ ] Yes
  * [X] No

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/main/steps/utils/fieldValidators.ts [hunk 1]
```diff
@@ -1,5 +1,18 @@
 export const EMOJI_PATTERN = /\p{Emoji_Presentation}|\p{Extended_Pictographic}|\u200D|\uFE0F/u;
 
+export const UNSAFE_TEXT_PATTERNS: RegExp[] = [
```
Comment (paddy-hmcts) [inline src/main/steps/utils/fieldValidators.ts]: I think perhaps it would be much better to use an npm library here that's been tried and tested:

https://www.npmjs.com/package/xss

Comment (sdmiddletonSolirius) [reply src/main/steps/utils/fieldValidators.ts]: Hi @paddy-hmcts - Have utilisted the xss package as discussed, with some modifications to allow legitimate use of the comparison operators.

File: src/main/assets/locales/cy/common.json [hunk 2]
```diff
@@ -48,6 +48,7 @@
     "defaultRequired": "cyThis field is required",
     "defaultInvalid": "cyInvalid format",
     "defaultMaxLength": "cyMust be {max} characters or fewer",
+    "defaultSpecialCharacter": "{fieldName} must only include letters a to z, and special characters such as hyphens, spaces and apostrophes",
```
Comment (andrewdami11) [inline src/main/assets/locales/cy/common.json:53]: Sibling Welsh keys use a `cy` prefix; this new `defaultSpecialCharacter` value is English. Should this use Welsh copy (or the same `cy…` placeholder pattern) for consistency?

File: src/test/unit/modules/steps/formBuilder/helpers.test.ts [hunk 3]
```diff
@@ -1517,6 +1517,105 @@ describe('formBuilder helpers', () => {
...
+                },
+              },
+              { value: 'notSure' },
+            ],
+          },
+        ];
+
+        const errors = validateForm(req, fields, {});
+
+        expect(errors['tenancyTypeConfirm.correctType']).toBeUndefined();
```
Comment (andrewdami11) [inline src/test/unit/modules/steps/formBuilder/helpers.test.ts:1583]: Could we add a test where a required field only contains _<script>…</script>_? After stripping it should be empty, and validation should still fail, no?

File: src/main/assets/locales/cy/common.json (hunk 2 above)
Comment (sdmiddletonSolirius) [reply src/main/assets/locales/cy/common.json:53]: Hey Andrew - the cy prefix is already built into the fieldName, so it's removed here as it was generating a double cycy on render - I learned this the hard way first time round.

File: src/test/unit/modules/steps/formBuilder/helpers.test.ts (hunk 3 above)
Comment (sdmiddletonSolirius) [reply src/test/unit/modules/steps/formBuilder/helpers.test.ts:1583]: Good catch - this has now been updated so inputs with only script tags will fail validation for being an empty response, before they were silently failing.

**Review by andrewdami11 — APPROVED** [review]

File: src/main/modules/steps/formBuilder/helpers.ts [hunk 4]
```diff
@@ -470,28 +486,31 @@ export function validateForm(
...
+          }
+
+          const translationLabelKey =
+            typeof field.translationKey === 'object' ? field.translationKey.label : field.translationKey;
+
+          const resolvedLabel = translationLabelKey && t ? getTranslation(t, translationLabelKey) : undefined;
+
+          const displayName = resolvedLabel ?? toSentenceCase(fieldName);
+
+          const defaultSpecialCharacterMsg = translations?.defaultSpecialCharacter?.replace('{fieldName}', displayName);
```
Comment (paddy-hmcts) [inline src/main/modules/steps/formBuilder/helpers.ts]: we shouldn't be doing this with translations, but using interpolation instead: https://www.i18next.com/translation-function/interpolation

File: src/main/modules/steps/formBuilder/helpers.ts [hunk 5]
```diff
@@ -470,28 +486,31 @@ export function validateForm(
...
+          errors[fieldName] =
+            defaultSpecialCharacterMsg ||
+            `${displayName} must only include letters a to z, and special characters such as hyphens, spaces and apostrophes`;
```
Comment (paddy-hmcts) [inline src/main/modules/steps/formBuilder/helpers.ts]: I don't think we should be providing default fallbacks for translation strings as the dev should know where it's broken and fix appropriately

File: src/main/modules/steps/formBuilder/helpers.ts [hunk 6]
```diff
@@ -470,28 +486,31 @@ export function validateForm(
...
-
-              const defaultSpecialCharacterMsg = translations?.defaultSpecialCharacter?.replace(
-                '{fieldName}',
-                displayName
-              );
-              errors[fieldName] =
-                defaultSpecialCharacterMsg ||
-                `${displayName} must only include letters a to z, and special characters such as hyphens, spaces and apostrophes`;
-            }
+            setDefaultSpecialCharacterError();
```
Comment (paddy-hmcts) [inline src/main/modules/steps/formBuilder/helpers.ts]: can I ask why this is actually required here? Isn't this part of a separate ticket that was already addressed? more for my own understanding

File: src/test/unit/app/controller/validation.test.ts [hunk 7]
```diff
@@ -104,5 +104,49 @@ describe('validateForm', () => {
       expect(errors.textareaField).toContain('must only include letters a to z');
       expect(errors.charCountField).toContain('must only include letters a to z');
     });
+
```
Comment (paddy-hmcts) [inline src/test/unit/app/controller/validation.test.ts:107]: can we please remove this before merging

**Review by paddy-hmcts — APPROVED** [review]
few minor comments but overall looks good

File: src/main/modules/steps/formBuilder/helpers.ts (hunk 6 above)
Comment (sdmiddletonSolirius) [reply src/main/modules/steps/formBuilder/helpers.ts]: Hi @paddy-hmcts - you are correct, this was originally part of the initial approach which was to reject illegal characters, and so we were piggy backing off the same error validation for emoji's - however, since adopting the xss sanitisation approach, I've now reverted this back so it doesn't encroach on the emoji implementation which is part of a separate ticket.

File: src/test/unit/app/controller/validation.test.ts (hunk 7 above)
Comment (sdmiddletonSolirius) [reply src/test/unit/app/controller/validation.test.ts:107]: Done

File: src/main/modules/steps/formBuilder/helpers.ts (hunk 4 above)
Comment (sdmiddletonSolirius) [reply src/main/modules/steps/formBuilder/helpers.ts]: Have now reverted to before XSS implementation

File: src/main/assets/locales/cy/common.json (hunk 2 above)
Comment (paddy-hmcts) [reply src/main/assets/locales/cy/common.json:53]: you probably also no longer need this key now

**Review by paddy-hmcts — APPROVED** [review]
